Improve reading full flash on capsule updates

SergiiDmytruk opened the pull request:
* Use a more rare (and no less applicable) error code to report an issue with reading current firmware.  This is should make failure easier to understand.
* Double the number of retries on calling into SMM to avoid getting `EFI_NO_RESPONSE`.

This is part of fixing https://github.com/Dasharo/dasharo-issues/issues/1338.

coreboot part: https://github.com/Dasharo/coreboot/pull/700

*ref: ncm-1850*

SergiiDmytruk committed:
DasharoPayloadPkg/FmpDeviceSmmLib: report EFI_END_OF_MEDIA if can't read current firmware

EFI_OUT_OF_RESOURCES was a bad choice as it most often suggests memory
allocation issue.  EFI_END_OF_MEDIA is not used and is therefore
unambiguous.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/SmmStoreLib: retry SMM 10 times instead of 5

At least Clevo laptops seem to be especially prone to SMIs being
occasionally ignored, so double the number of tries.

Practically speaking, retrying 6 times would probably have the same
effect, but there should be no negative side-effects of 10 retries as it
affects only a highly unlikely cases of multiple failures in a row.

The approximate failure rate with 5 retries was ~1/128000 SMM requests.
This change should lower the rate by several orders of magnitude.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

philipanda reviewed (commented):
I see no issues with these changes
*ediit (misslick, I've intended to approve)

philipanda reviewed (approved):
I see no issues with these changes

mkopec reviewed (approved):
A bit surprising that 5 retries weren't enough, when on alder lake even just 1 retry was sufficient

SergiiDmytruk commented:
https://github.com/Dasharo/dasharo-issues/issues/1338#issuecomment-2996002475 shows this isn't enough to fix the issue, so let's not merge it yet in case further EDK modifications are needed.

SergiiDmytruk committed:
DasharoPayloadPkg/SmmStoreLib: log unexpected response from SMM

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
DasharoPayloadPkg/SmmStoreLib: add NOPs to TriggerSmi

There is a chance that adding NOPs has the effect of increasing the
probability of success.  The reasoning is that SMI might not happen if
an interrupt occurs due to IRET in the handler suppressing the SMI, thus
extra instructions could give the interrupt a higher chance of being
processed before CMP instruction is run.

From Intel SDM 3c:

  An SMI may be blocked for one instruction following execution of STI,
  MOV to SS, or POP into SS.

From Intel SDM 2a:

  If the return is to another privilege level, the IRET instruction also
  pops the stack pointer and SS from the stack, before resuming program
  execution.

So it's not entirely clear this could happen in the firmware running at
ring 0, but it won't hurt either.

Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

mkopec commented:
The double `nop`s in the SMM calling code also help with https://github.com/Dasharo/dasharo-issues/issues/1349 or at least lower the reproduction rate somewhat.

krystian-hebel commented on DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c:
```diff
@@ -981,7 +981,7 @@ FmpDeviceSetImageWithStatus (
       __FUNCTION__
       ));
-    return EFI_OUT_OF_RESOURCES;
+    return EFI_END_OF_MEDIA;
```
Commit title still mentions `EFI_OUT_OF_RESOURCES`.

SergiiDmytruk replied:
https://github.com/Dasharo/edk2/pull/255/commits/8f29b71f2b5205bdeb4952463f506f94390558c4
